### PR TITLE
separate mutex impl, fix win test, rec mutex works

### PR DESCRIPTION
**Pull request by earonesty**
- separate out the recursive shared mutex code
- counts separately maintained

### CODE REVIEW

File: src/hilok.hpp [hunk 1]
```diff
@@ -214,10 +159,6 @@ class HiHandle {
     HiHandle(std::shared_ptr<HiLok> mgr, bool shared, std::shared_ptr<HiKeyNode> ref) :
         m_shared(shared), m_ref(ref), m_mgr(mgr), m_released(false) {
     }
-    HiHandle ( HiHandle && ) = default;
```
Comment (colatkinson) [inline src/hilok.hpp:217]: Why was this removed?

File: tests/test.cpp [hunk 2]
```diff
@@ -188,6 +195,99 @@ TEST_CASE( "rlock-thread", "[basic]" ) {
...
+    auto dur = std::chrono::duration<double>(end - start);
+    CHECK(dur.count() >= 0.01);
+    }
+    
+    thread_lock->release();
+    thread.join();
+}
+
+
+TEST_CASE( "timed-hlock-rec", "[basic]" ) {
```
Comment (colatkinson) [inline tests/test.cpp]: nit: Does catch2 support doing assertions in separate funcs, or any kind of parameterization? If so, would be nice to dedupe this and the previous test since they're almost identical

**Review by colatkinson — APPROVED** [review]

File: tests/test.cpp (hunk 2 above)
Comment (earonesty) [reply tests/test.cpp]: maybe?  i can try

File: src/hilok.hpp (hunk 1 above)
Comment (earonesty) [reply src/hilok.hpp:217]: accidentally

Comment (earonesty) [reply src/hilok.hpp:217]: not sure, merge issuse maybe.

File: tests/test.cpp (hunk 2 above)
Comment (earonesty) [reply tests/test.cpp]: yeah, works fine
